internal/periodic_checker/server.go

```diff
@@ -78,48 +54,54 @@ func (s *Server) Run(ctx context.Context) error {
 
-	// resource sync
-	resourceSync := tasks.NewResourceSync(callbackManager, serviceHandler, s.log, s.cfg.GitOps.IgnoreResourceUpdates)
-	resourceSyncThread := s.newTaskThread(ctx, tasks.ResourceSyncTaskName, "ResourceSync", 2*time.Minute, resourceSync.Poll)
-	resourceSyncThread.Start()
-	defer resourceSyncThread.Stop()
-
-	// device disconnected
-	deviceDisconnected := tasks.NewDeviceDisconnected(s.log, serviceHandler)
-	deviceDisconnectedThread := s.newTaskThread(ctx, tasks.DeviceDisconnectedTaskName, "Device disconnected", tasks.DeviceDisconnectedPollingInterval, deviceDisconnected.Poll)
-	deviceDisconnectedThread.Start()
-	defer deviceDisconnectedThread.Stop()
-
-	// Rollout device selection
-	rolloutDeviceSelection := device_selection.NewReconciler(serviceHandler, callbackManager, s.log)
-	rolloutDeviceSelectionThread := s.newTaskThread(ctx, device_selection.DeviceSelectionTaskName, "Rollout device selection", device_selection.RolloutDeviceSelectionInterval, rolloutDeviceSelection.Reconcile)
-	rolloutDeviceSelectionThread.Start()
-	defer rolloutDeviceSelectionThread.Stop()
-
-	// Rollout disruption budget
-	disruptionBudget := disruption_budget.NewReconciler(serviceHandler, callbackManager, s.log)
-	disruptionBudgetThread := s.newTaskThread(ctx, disruption_budget.DisruptionBudgetTaskName, "Disruption budget", disruption_budget.DisruptionBudgetReconcilationInterval, disruptionBudget.Reconcile)
-	disruptionBudgetThread.Start()
-	defer disruptionBudgetThread.Stop()
-
-	// Event cleanup
-	eventCleanup := tasks.NewEventCleanup(s.log, serviceHandler, s.cfg.Service.EventRetentionPeriod)
-	eventCleanupThread := s.newTaskThread(ctx, tasks.EventCleanupTaskName, "Event cleanup", tasks.EventCleanupPollingInterval, eventCleanup.Poll)
-	eventCleanupThread.Start()
-	defer eventCleanupThread.Stop()
+	// Create channel manager for task distribution
+	channelManagerConfig := ChannelManagerConfig{
+		Log: s.log,
+	}
+	if s.cfg.Periodic != nil {
+		channelManagerConfig.ChannelBufferSize = s.cfg.Periodic.Consumers * 2
+	}
+	channelManager := NewChannelManager(channelManagerConfig)
+	defer channelManager.Close()
+
+	// Periodic task consumer
+	consumerConfig := PeriodicTaskConsumerConfig{
+		ChannelManager: channelManager,
+		Log:            s.log,
+		Executors:      periodicTaskExecutors,
+	}
+	if s.cfg.Periodic != nil {
+		consumerConfig.ConsumerCount = s.cfg.Periodic.Consumers
+	}
+	periodicTaskConsumer := NewPeriodicTaskConsumer(consumerConfig)
+	if err := periodicTaskConsumer.Start(ctx); err != nil {
+		return err
+	}
+	defer periodicTaskConsumer.Stop()
+
+	// Periodic task publisher
+	publisherConfig := PeriodicTaskPublisherConfig{
+		Log:            s.log,
+		OrgService:     serviceHandler,
+		TasksMetadata:  periodicTasks,
+		ChannelManager: channelManager,
+	}
+	periodicTaskPublisher, err := NewPeriodicTaskPublisher(publisherConfig)
+	if err != nil {
+		return err
+	}
+	periodicTaskPublisher.Start(ctx)
+	defer periodicTaskPublisher.Stop()
```

Comment:
This I think will just hang since you are dependent on the context to be cancelled for the wait group in this method to be done, but cancel won't be called until this defer returns (defers are processed in reverse order).